Propagate errors for saveTargetAudience

jorgemd24 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

This is a follow-up of https://github.com/woocommerce/google-listings-and-ads/issues/1331#issuecomment-1109566842 & https://github.com/woocommerce/google-listings-and-ads/issues/1331#issuecomment-1110084066

This PR removes the `try` and `catch` from `saveTargetAudience`, allowing the error to bubble up and be handled by the consumer.  In the following example:

https://github.com/woocommerce/google-listings-and-ads/blob/189a55a29a368b0a2ce9150deb3ca70c6077de29/js/src/edit-free-campaign/index.js#L203-L208

if `saveTargetAudience` failed, the consumer was not able to tell if `saveTargetAudience` failed because the error was already caught. Using this PR, `saveTargetAudience` will be shown as `rejected`.

For more context see issue #1331.

I found only two places where `saveTargetAudience` is called:

1.  https://github.com/woocommerce/google-listings-and-ads/blob/189a55a29a368b0a2ce9150deb3ca70c6077de29/js/src/edit-free-campaign/index.js#L204 (This error will be handle by a custom hook, see https://github.com/woocommerce/google-listings-and-ads/issues/1331#issuecomment-1109566842

2. https://github.com/woocommerce/google-listings-and-ads/blob/a2e05d7361a86efb33dd6dc2d86053ba62230039/js/src/setup-mc/setup-stepper/choose-audience/useAutoSaveTargetAudienceEffect.js#L26-L27  This PR updates the useAutoSaveTargetAudienceEffect to catch the error.

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Go to the onboarding page step 2.

![image](https://user-images.githubusercontent.com/2488994/177545042-1cf4421f-5460-4e29-b4d8-54be23fb7a6d.png)

2. Update the countries, no errors should be shown.
3. If you provoke an error, a notice will be displayed. For example, setting an incorrect URL in the following line: https://github.com/woocommerce/google-listings-and-ads/blob/a892de55a797a0feaf20e2d7b88ad25755614417/js/src/data/actions.js#L725 

![image](https://user-images.githubusercontent.com/2488994/177545337-003976fa-60d3-45cb-af70-1bc973fab0c2.png)


### Additional details:
- The next step will be to remove the try and catch from `saveSettings` so we can handle the errors in situations like the following:

```
const results = await Promise.allSettled( [
	saveTargetAudience( targetAudience ), // remove its try/catch in actions.js. (This PR)
	saveSettings( settings ), // TODO: remove its try/catch in actions.js.
	saveShippingRates( shippingRates ), // already throws error when it fails.
	saveShippingTimes( shippingTimes ), // already throws error when it fails.
] );
```
 <!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

>Tweak - Remove try and catch in saveTargetAudience action.

jorgemd24 committed:
Remove try and catch from saveTargetAudience

jorgemd24 committed:
Adjust useAutoSaveTargetAudienceEffect to catch the error

jorgemd24 committed:
Fix typo saveTargetAudience

puntope commented:
I was testing this and I couldn't render the error notice in Edit Free Listings.

I provoke a 404 error like this editing `saveTargetAudience` in `actions.js:`

```
yield apiFetch( {
		path: `${ API_NAMESPACE }/mc/target_audiencesssss`,
		method: 'POST',
		data: targetAudience,
	} );
```

Then I still see the success message

<img width="1693" alt="Screenshot 2022-07-07 at 18 17 56" src="https://user-images.githubusercontent.com/5908855/177796069-9415ee70-82b0-4f28-8276-989393445ac3.png">

Notice in Onboarding I see it working perfectly.

ecgan commented:
@puntope , I believe this PR covers only the Setup MC onboarding flow. The Edit Free Listings flow will be covered / handled by a custom hook later.

jorgemd24 commented:
Thanks, @puntope, for reviewing this.

> I was testing this and I couldn't render the error notice in Edit Free Listings.

That's expected, as it is mentioned on this https://github.com/woocommerce/google-listings-and-ads/issues/1331#issuecomment-1110084066, the idea was to split the changes in different PRs to minimize the risks: first to propagate the errors for the `saveTargetAudience`, the same with `saveSettings` and finally create the custom hook to handle cases such as `Edit Free Listings` where multiple promises are called with `allSettled`. Currently, If `saveSettings` fails it would still show the success message independently of `saveTargetAudience`. 

I am happy to discuss if it is convenient to add the custom hook in this PR, however, we need to keep in mind some requirements that could make this PR a bit more complex and could divert from its original intention which is to propagate the errors of `saveTargetAudience`:

- The error notice needs to be able to show multiple errors and be coherent with i18n. See the second part of this https://github.com/woocommerce/google-listings-and-ads/issues/1331#issuecomment-1105050752 for more context.
- In the case that we would like to show multiple errors as a bullet list, we will need to use a different type of notice. As far as I know, the `createNotice` only allows string content and single-line.

```
There are errors in the following actions: 
 - There was an error trying to save the target audience
 - There was an error trying to save settings
```

- When evaluating the results from `allSettled` we will need to match it with its user-friendly error.

The hook could look something like the code below, however we will need to find a way to make it more coherent with i18n and more generic:

```
const useCustomAllSettled = () => {
	const { createNotice } = useDispatchCoreNotices();

	return async ( promises, errorsText ) => {

		const results = await Promise.allSettled( promises );

		const rejected = results.reduce( ( acc, item, index ) => {
			if ( item.status === 'rejected' && errorsText[ index ] ) {
				acc.push( errorsText[ index ] );
			}
			return acc;
		}, [] );

		const hasErrors = !! rejected.length;

		if ( hasErrors ) {
			createNotice(
				'error',
				`There were errors trying to save: ${ rejected }. Other changes have been saved. Please try again later.`
			);
		}

		return hasErrors;
	};
};
```


 Let me know what you think!

puntope commented:
> Let me know what you think!

We can handle it in separate PR no problem. I wan't sure if we were covering those as well.

puntope commented:
> * In the case that we would like to show multiple errors as a bullet list, we will need to use a different type of notice. As far as I know, the `createNotice` only allows string content and single-line.

I like the idea of concat errors like ... actionA, actionB, actionC as you defined them in the hook.

```There were errors trying to save: target audience, taxes and budget. Other changes have been saved. Please try again later.``` 

I personally hate when some APP show me "try again later..." and also I guess we should add an "and" in the last item somehow. But I guess the copy is more a question for design team. The idea looks good for me. 

Ok also to have it in other PR

puntope commented:
One thing I saw is with a fresh onboarding, in case we force the error I see the message but I can still continue to next step, however, I just saw the spinner running forever. 

Then I removed the error, completed step 2 and the problem is not happening anymore even if I force again the error in target audience. 

https://user-images.githubusercontent.com/5908855/177945394-3dd0c94e-ad91-4eac-a388-5f9aaf26255f.mov

Wondering if we should block the continue button in case of error. But this could also be done/discussed in further PR

puntope commented:
💅  Wondering if we could extend our tests adding some checks...  for example checking that no exception is thrown by the action.

jorgemd24 committed:
Disable continue button if autosave target audience fails

jorgemd24 committed:
Add tests useAutoSaveTargetAudienceEffect

jorgemd24 commented:
@puntope,

>Wondering if we should block the continue button in case of error. But this could also be done/discussed in further PR

I have added it in this PR, the continue button is disabled if the autosaving fails. See video:

https://user-images.githubusercontent.com/2488994/178262677-2d8f0483-871f-4439-bca2-9baef7544713.mp4

>Wondering if we could extend our tests adding some checks... for example checking that no exception is thrown by the action.

The action throws an exception and it is caught by the consumer, (the autosaving hook in this case ). I added some tests for the autosaving hook. 

@puntope, I guess this is ready for another round! Thanks for your comments!

puntope commented on js/src/setup-mc/setup-stepper/choose-audience/form-content.js:
```diff
@@ -32,8 +32,12 @@ const FormContent = ( props ) => {
+	const autoSaveCallback = ( resultAutoSave ) =>
+		setAutoSaved( resultAutoSave );
+
+	useAutoSaveTargetAudienceEffect( values, autoSaveCallback );
```
💅 

`useAutoSaveTargetAudienceEffect( values, setAutoSaved );` 

Also works ☺️

puntope reviewed (approved):
✅  LGTM

Thanks for the changes, tweaks and tests 👨‍💻

jorgemd24 committed:
simplify autoSaveCallback Target Audience

jorgemd24 commented:
Thanks, @puntope for reviewing this and your suggestions 💯
